File: Semester2/Graphs/homeworks/lab1/c++/tests/graph_tests.cpp
#include "gtest/gtest.h"
#include "Graph.h"

using namespace std;

class GraphTest:public testing::Test{
protected:
    Graph graph;
    void SetUp() override {
        vector<string> vertices = {"1", "2", "3"};
        vector<tuple<string, string, int> > edges = {{"1", "2", 0}, {"2", "1", 10}, {"1", "3", 3}};
        graph = Graph(vertices, edges);
    }
};

TEST_F(GraphTest, test_constructor){
    ASSERT_THROW(graph = Graph({"1", "2", "3"}, {{"1", "2", 0}, {"2", "1", 10}, {"1", "2", 3}}), GraphException);
    ASSERT_THROW(graph = Graph({"1", "2", "3"}, {{"1", "2", 0}, {"2", "1", 10}, {"1", "4", 3}}), GraphException);
    graph = Graph({"1", "2", "3"}, {{"1", "2", 0}, {"2", "1", 10}, {"1", "3", 3}});
}

TEST_F(GraphTest, test_parse_vertices){
    vector<string> vertices = {"1", "2", "3"};
    ASSERT_EQ(this->graph.parse_vertices(), vertices);
}

TEST_F(GraphTest, test_is_edge){
    vector<string> vertices = {"1", "2", "3"};
    ASSERT_THROW(this->graph.is_edge("4", "2"), GraphException);
    ASSERT_EQ(this->graph.is_edge("1", "1"), false);
    ASSERT_EQ(this->graph.is_edge("1", "2"), true);
    ASSERT_EQ(this->graph.is_edge("1", "3"), true);
    ASSERT_EQ(this->graph.is_edge("2", "1"), true);
    ASSERT_EQ(this->graph.is_edge("2", "2"), false);
    ASSERT_EQ(this->graph.is_edge("2", "3"), false);
    ASSERT_EQ(this->graph.is_edge("3", "1"), false);
    ASSERT_EQ(this->graph.is_edge("3", "2"), false);
    ASSERT_EQ(this->graph.is_edge("3", "3"), false);
}

TEST_F(GraphTest, test_get_in_degree){
    ASSERT_THROW(this->graph.get_in_degree("4"), GraphException);
    ASSERT_EQ(this->graph.get_in_degree("1"), 1);
    ASSERT_EQ(this->graph.get_in_degree("2"), 1);
    ASSERT_EQ(this->graph.get_in_degree("3"), 1);
}

TEST_F(GraphTest, test_get_out_degree){
    ASSERT_THROW(this->graph.get_out_degree("4"), GraphException);
    ASSERT_EQ(this->graph.get_out_degree("1"), 2);
    ASSERT_EQ(this->graph.get_out_degree("2"), 1);
    ASSERT_EQ(this->graph.get_out_degree("3"), 0);
}

TEST_F(GraphTest, test_parse_outbound_edges){
    ASSERT_THROW(this->graph.parse_outbound_edges("4"), GraphException);
    vector<string> vertices;
    vertices = {"2", "3"};
    ASSERT_EQ(this->graph.parse_outbound_edges("1"), vertices);
    vertices = {"1"};
    ASSERT_EQ(this->graph.parse_outbound_edges("2"), vertices);
    vertices = {};
    ASSERT_EQ(this->graph.parse_outbound_edges("3"), vertices);
}

TEST_F(GraphTest, test_parse_inbound_edges){
    ASSERT_THROW(this->graph.parse_inbound_edges("4"), GraphException);
    vector<string> vertices;
    vertices = {"2"};
    ASSERT_EQ(this->graph.parse_inbound_edges("1"), vertices);
    vertices = {"1"};
    ASSERT_EQ(this->graph.parse_inbound_edges("2"), vertices);
    vertices = {"1"};
    ASSERT_EQ(this->graph.parse_inbound_edges("3"), vertices);
}

TEST_F(GraphTest, test_get_edge_cost){
    ASSERT_THROW(this->graph.get_edge_cost("4", "1"), GraphException);
    ASSERT_THROW(this->graph.get_edge_cost("3", "2"), GraphException);
    ASSERT_THROW(this->graph.get_edge_cost("1", "4"), GraphException);
    ASSERT_EQ(this->graph.get_edge_cost("1", "2"), 0);
    ASSERT_EQ(this->graph.get_edge_cost("1", "3"), 3);
    ASSERT_EQ(this->graph.get_edge_cost("2", "1"), 10);
}

TEST_F(GraphTest, test_modify_edge_cost){
    ASSERT_THROW(this->graph.modify_edge_cost("4", "1", 0), GraphException);
    ASSERT_THROW(this->graph.modify_edge_cost("3", "2", 0), GraphException);
    ASSERT_THROW(this->graph.modify_edge_cost("1", "4", 0), GraphException);
    this->graph.modify_edge_cost("2", "1", -7);
    ASSERT_EQ(this->graph.get_edge_cost("1", "2"), 0);
    ASSERT_EQ(this->graph.get_edge_cost("1", "3"), 3);
    ASSERT_EQ(this->graph.get_edge_cost("2", "1"), -7);
}

TEST_F(GraphTest, test_copy){
    Graph graph2 = Graph(this->graph);
    this->graph.modify_edge_cost("2", "1", -7);
    ASSERT_EQ(graph2.get_edge_cost("1", "2"), 0);
    ASSERT_EQ(graph2.get_edge_cost("2", "1"), 10);
    ASSERT_EQ(graph2.get_edge_cost("1", "3"), 3);
    vector<string> vertices;
    vertices = {"2", "3"};
    ASSERT_EQ(graph2.parse_outbound_edges("1"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph2.parse_outbound_edges("2"), vertices);
    vertices = {};
    ASSERT_EQ(graph2.parse_outbound_edges("3"), vertices);
    vertices = {"2"};
    ASSERT_EQ(graph2.parse_inbound_edges("1"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph2.parse_inbound_edges("2"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph2.parse_inbound_edges("3"), vertices);
    vertices = {"1", "2", "3"};
    ASSERT_EQ(graph2.parse_vertices(), vertices);
}

TEST_F(GraphTest, test_add_vertex){
    ASSERT_THROW(this->graph.add_vertex("1"), GraphException);
    this->graph.add_vertex("10");
    this->graph.add_vertex("4");
    ASSERT_EQ(graph.get_edge_cost("1", "2"), 0);
    ASSERT_EQ(graph.get_edge_cost("2", "1"), 10);
    ASSERT_EQ(graph.get_edge_cost("1", "3"), 3);
    vector<string> vertices;
    vertices = {"2", "3"};
    ASSERT_EQ(graph.parse_outbound_edges("1"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph.parse_outbound_edges("2"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_outbound_edges("3"), vertices);
    vertices = {"2"};
    ASSERT_EQ(graph.parse_inbound_edges("1"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph.parse_inbound_edges("2"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph.parse_inbound_edges("3"), vertices);
    vertices = {"1", "10", "2", "3", "4"};
    ASSERT_EQ(graph.parse_vertices(), vertices);
}

TEST_F(GraphTest, test_remove_vertex){
    ASSERT_THROW(this->graph.remove_vertex("10"), GraphException);
    this->graph.remove_vertex("2");
    ASSERT_EQ(graph.get_edge_cost("1", "3"), 3);
    vector<string> vertices;
    vertices = {"3"};
    ASSERT_EQ(graph.parse_outbound_edges("1"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_outbound_edges("3"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_inbound_edges("1"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph.parse_inbound_edges("3"), vertices);
    vertices = {"1", "3"};
    ASSERT_EQ(graph.parse_vertices(), vertices);
}

TEST_F(GraphTest, test_add_edge){
    ASSERT_THROW(this->graph.add_edge("3", "4", 10), GraphException);
    ASSERT_THROW(this->graph.add_edge("4", "3", 10), GraphException);
    this->graph.add_vertex("10");
    this->graph.add_vertex("4");
    this->graph.add_edge("10", "2", 1);
    this->graph.add_edge("10", "1", 10);
    this->graph.add_edge("1", "10", 3);
    this->graph.add_edge("2", "3", 4);

    ASSERT_EQ(this->graph.get_edge_cost("1", "3"), 3);
    ASSERT_EQ(this->graph.get_edge_cost("10", "2"), 1);
    ASSERT_EQ(this->graph.get_edge_cost("10", "1"), 10);
    ASSERT_EQ(this->graph.get_edge_cost("1", "10"), 3);
    ASSERT_EQ(this->graph.get_edge_cost("2", "3"), 4);


    vector<string> vertices;
    vertices = {"2", "3", "10"};
    ASSERT_EQ(graph.parse_outbound_edges("1"), vertices);
    vertices = {"1", "3"};
    ASSERT_EQ(graph.parse_outbound_edges("2"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_outbound_edges("3"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_outbound_edges("4"), vertices);
    vertices = {"2", "1"};
    ASSERT_EQ(graph.parse_outbound_edges("10"), vertices);
    vertices = {"2", "10"};
    ASSERT_EQ(graph.parse_inbound_edges("1"), vertices);
    vertices = {"1", "10"};
    ASSERT_EQ(graph.parse_inbound_edges("2"), vertices);
    vertices = {"1", "2"};
    ASSERT_EQ(graph.parse_inbound_edges("3"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph.parse_inbound_edges("10"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_inbound_edges("4"), vertices);
    vertices = {"1", "10", "2", "3", "4"};
    ASSERT_EQ(graph.parse_vertices(), vertices);
}

TEST_F(GraphTest, test_remove_edge){
    ASSERT_THROW(this->graph.remove_edge("3", "4"), GraphException);
    ASSERT_THROW(this->graph.remove_edge("3", "2"), GraphException);
    ASSERT_THROW(this->graph.remove_edge("4", "3"), GraphException);

    this->graph.add_vertex("10");
    this->graph.add_vertex("4");
    this->graph.add_edge("10", "2", 1);
    this->graph.add_edge("10", "1", 10);
    this->graph.remove_edge("1", "2");
    this->graph.remove_edge("2", "1");
    this->graph.add_edge("1", "10", 3);
    this->graph.add_edge("2", "3", 4);

    ASSERT_EQ(this->graph.get_edge_cost("1", "3"), 3);
    ASSERT_EQ(this->graph.get_edge_cost("10", "2"), 1);
    ASSERT_EQ(this->graph.get_edge_cost("10", "1"), 10);
    ASSERT_EQ(this->graph.get_edge_cost("1", "10"), 3);
    ASSERT_EQ(this->graph.get_edge_cost("2", "3"), 4);


    vector<string> vertices;
    vertices = {"3", "10"};
    ASSERT_EQ(graph.parse_outbound_edges("1"), vertices);
    vertices = {"3"};
    ASSERT_EQ(graph.parse_outbound_edges("2"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_outbound_edges("3"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_outbound_edges("4"), vertices);
    vertices = {"2", "1"};
    ASSERT_EQ(graph.parse_outbound_edges("10"), vertices);
    vertices = {"10"};
    ASSERT_EQ(graph.parse_inbound_edges("1"), vertices);
    vertices = {"10"};
    ASSERT_EQ(graph.parse_inbound_edges("2"), vertices);
    vertices = {"1", "2"};
    ASSERT_EQ(graph.parse_inbound_edges("3"), vertices);
    vertices = {"1"};
    ASSERT_EQ(graph.parse_inbound_edges("10"), vertices);
    vertices = {};
    ASSERT_EQ(graph.parse_inbound_edges("4"), vertices);
    vertices = {"1", "10", "2", "3", "4"};
    ASSERT_EQ(graph.parse_vertices(), vertices);
}

TEST_F(GraphTest, test_eq){
    ASSERT_EQ(Graph(this->graph), this->graph);
    Graph graph2 = Graph(this->graph);
    graph2.add_vertex("4");
    ASSERT_NE(graph2, this->graph);
    graph2.remove_vertex("1");
    ASSERT_NE(graph2, this->graph);
    graph2.remove_vertex("4");
    ASSERT_NE(graph2, this->graph);
    graph2.add_vertex("1");
    ASSERT_NE(graph2, this->graph);
    graph2.add_edge("1", "2", 0);
    ASSERT_NE(graph2, this->graph);
    graph2.add_edge("1", "3", 3);
    ASSERT_NE(graph2, this->graph);
    graph2.add_edge("2", "1", 10);
    ASSERT_EQ(graph2, this->graph);
    this->graph.add_edge("1", "1", 0);
    graph2.add_edge("1", "1", 10);
    ASSERT_NE(graph2, this->graph);
}

TEST_F(GraphTest, test_read_graph){

    ASSERT_THROW(read_graph("aaaa"), GraphException);
    ASSERT_THROW(read_graph("aaaa.modified"), GraphException);
    ASSERT_THROW(read_graph("../../tests/test_files/fail1.txt"), GraphException);
    ASSERT_THROW(read_graph("../../tests/test_files/fail2.txt"), GraphException);
    ASSERT_THROW(read_graph("../../tests/test_files/fail3.txt"), GraphException);
    ASSERT_THROW(read_graph("../../tests/test_files/fail1.modified.txt"), GraphException);
    ASSERT_THROW(read_graph("../../tests/test_files/fail2.modified.txt"), GraphException);
    ASSERT_THROW(read_graph("../../tests/test_files/fail3.modified.txt"), GraphException);
    ASSERT_THROW(read_graph("../../tests/test_files/fail4.modified.txt"), GraphException);
    vector<string> vertices = {"0", "1", "2", "3", "4"};
    vector<tuple<string, string, int> > edges;
    edges = {
            {"0", "0", 1},
            {"0", "1", 7},
            {"1", "2", 2},
            {"2", "1", -1},
            {"1", "3", 8},
            {"2", "3", 5}
    };
    Graph graph1 = read_graph("../../tests/test_files/graph1.txt");
    ASSERT_EQ(graph1, Graph(vertices, edges));
    Graph graph2 = read_graph("../../tests/test_files/graph2.modified.txt");
    vertices = {
            "1", "5", "2", "0", "3"
    };
    edges = {
            {"1", "5", 10},
            {"1", "2", 0},
            {"0", "3", 4},
            {"3", "5", 20},
            {"2", "1", 4},
            {"1", "1", 5}
    };
    ASSERT_EQ(graph2, Graph(vertices, edges));
}

TEST_F(GraphTest, test_write_graph){

    ASSERT_THROW(write_graph("aaaa",this->graph), GraphException);
    ASSERT_THROW(write_graph("aaaa.modified",this->graph), GraphException);
    write_graph("./aaaa.modified.txt", this->graph);
    ASSERT_EQ(read_graph("./aaaa.modified.txt"), this->graph);
}

TEST_F(GraphTest, test_random_graph){

    ASSERT_THROW(random_graph(-10, 2), GraphException);
    ASSERT_THROW(random_graph(10, -2), GraphException);
    ASSERT_THROW(random_graph(6, 40), GraphException);
    ASSERT_THROW(write_graph("aaaa.modified",this->graph), GraphException);
    Graph graph = random_graph(6, 3);
    ASSERT_EQ(graph.parse_vertices().size(), 6);
    int cnt_edges = 0;
    for(auto it:graph.parse_vertices()){
        cnt_edges += graph.get_out_degree(it);
    }
    ASSERT_EQ(cnt_edges, 3);
}

File: Semester2/Graphs/homeworks/lab3/python/GraphTests.py
from Graph import Graph, read_graph, write_graph, random_graph, floyd_warshall
from math import inf
import unittest


class GraphTests(unittest.TestCase):

    def setUp(self):
        self.graph = Graph(["1", "2", "3"], [("1", "2", 0), ("2", "1", 10), ("1", "3", 3)])

    def test_constructor(self):
        with self.assertRaises(Exception):
            Graph(1, [])
        with self.assertRaises(Exception):
            Graph([], "")
        with self.assertRaises(Exception):
            graph = Graph([1, 2, 3], [(1, 2, 0)])

        with self.assertRaises(Exception):
            graph = Graph(["1", "2", "3"], [(1, 2, 0)])

        with self.assertRaises(Exception):
            graph = Graph(["1", "2", "3"], [("1", "2", "0")])

        with self.assertRaises(Exception):
            graph = Graph(["1", "2", "3"], [("1", "2", 0), ("2", "1", 10), ("1", "2", 3)])
        graph = Graph(["1", "2", "3"], [("1", "2", 0), ("2", "1", 10), ("1", "3", 3)])

    def test_parse_vertices(self):
        self.assertEqual(self.graph.parse_vertices(), ["1", "2", "3"])

    def test_is_edge(self):
        with self.assertRaises(Exception):
            self.graph.is_edge(1, 2)
        self.assertEqual(self.graph.is_edge("1", "1"), False)
        self.assertEqual(self.graph.is_edge("1", "2"), True)
        self.assertEqual(self.graph.is_edge("1", "3"), True)
        self.assertEqual(self.graph.is_edge("2", "1"), True)
        self.assertEqual(self.graph.is_edge("2", "2"), False)
        self.assertEqual(self.graph.is_edge("2", "3"), False)
        self.assertEqual(self.graph.is_edge("3", "1"), False)
        self.assertEqual(self.graph.is_edge("3", "2"), False)
        self.assertEqual(self.graph.is_edge("3", "3"), False)

    def test_get_in_degree(self):
        with self.assertRaises(Exception):
            self.graph.get_in_degree(10)
        with self.assertRaises(Exception):
            self.graph.get_in_degree("10")
        self.assertEqual(self.graph.get_in_degree("1"), 1)
        self.assertEqual(self.graph.get_in_degree("2"), 1)
        self.assertEqual(self.graph.get_in_degree("3"), 1)

    def test_get_out_degree(self):
        with self.assertRaises(Exception):
            self.graph.get_out_degree(10)
        with self.assertRaises(Exception):
            self.graph.get_out_degree("10")
        self.assertEqual(self.graph.get_out_degree("1"), 2)
        self.assertEqual(self.graph.get_out_degree("2"), 1)
        self.assertEqual(self.graph.get_out_degree("3"), 0)

    def test_parse_outbound_edges(self):
        with self.assertRaises(Exception):
            self.graph.parse_outbound_edges(10)
        with self.assertRaises(Exception):
            self.graph.parse_outbound_edges("4")
        self.assertEqual(self.graph.parse_outbound_edges("1"), ["2", "3"])
        self.assertEqual(self.graph.parse_outbound_edges("2"), ["1"])
        self.assertEqual(self.graph.parse_outbound_edges("3"), [])

    def test_parse_inbound_edges(self):
        with self.assertRaises(Exception):
            self.graph.parse_inbound_edges(10)
        with self.assertRaises(Exception):
            self.graph.parse_inbound_edges("4")

        self.assertEqual(self.graph.parse_inbound_edges("1"), ["2"])
        self.assertEqual(self.graph.parse_inbound_edges("2"), ["1"])
        self.assertEqual(self.graph.parse_inbound_edges("3"), ["1"])

    def test_get_edge_cost(self):
        with self.assertRaises(Exception):
            self.graph.get_edge_cost(3, 1)
        with self.assertRaises(Exception):
            self.graph.get_edge_cost("3", "1")
        self.assertEqual(self.graph.get_edge_cost("1", "2"), 0)
        self.assertEqual(self.graph.get_edge_cost("2", "1"), 10)
        self.assertEqual(self.graph.get_edge_cost("1", "3"), 3)

    def test_modify_edge_cost(self):
        with self.assertRaises(Exception):
            self.graph.modify_edge_cost(3, 1, 10)
        with self.assertRaises(Exception):
            self.graph.modify_edge_cost("3", "1", 10)
        with self.assertRaises(Exception):
            self.graph.modify_edge_cost("2", "1", "10")
        self.graph.modify_edge_cost("2", "1", -7)
        self.assertEqual(self.graph.get_edge_cost("1", "2"), 0)
        self.assertEqual(self.graph.get_edge_cost("2", "1"), -7)
        self.assertEqual(self.graph.get_edge_cost("1", "3"), 3)

    def test_copy(self):
        graph2 = self.graph.copy()
        self.graph.modify_edge_cost("2", "1", -7)
        self.assertEqual(graph2.get_edge_cost("1", "2"), 0)
        self.assertEqual(graph2.get_edge_cost("2", "1"), 10)
        self.assertEqual(graph2.get_edge_cost("1", "3"), 3)
        self.assertEqual(graph2.parse_outbound_edges("1"), ["2", "3"])
        self.assertEqual(graph2.parse_outbound_edges("2"), ["1"])
        self.assertEqual(graph2.parse_outbound_edges("3"), [])
        self.assertEqual(graph2.parse_inbound_edges("1"), ["2"])
        self.assertEqual(graph2.parse_inbound_edges("2"), ["1"])
        self.assertEqual(graph2.parse_inbound_edges("3"), ["1"])
        self.assertEqual(graph2.parse_vertices(), ["1", "2", "3"])

    def test_add_vertex(self):
        with self.assertRaises(Exception):
            self.graph.add_vertex(1)
        with self.assertRaises(Exception):
            self.graph.add_vertex("1")
        self.graph.add_vertex("4")
        self.graph.add_vertex("10")
        self.assertEqual(self.graph.get_edge_cost("1", "2"), 0)
        self.assertEqual(self.graph.get_edge_cost("2", "1"), 10)
        self.assertEqual(self.graph.get_edge_cost("1", "3"), 3)
        self.assertEqual(self.graph.parse_outbound_edges("1"), ["2", "3"])
        self.assertEqual(self.graph.parse_outbound_edges("2"), ["1"])
        self.assertEqual(self.graph.parse_outbound_edges("3"), [])
        self.assertEqual(self.graph.parse_inbound_edges("1"), ["2"])
        self.assertEqual(self.graph.parse_inbound_edges("2"), ["1"])
        self.assertEqual(self.graph.parse_inbound_edges("3"), ["1"])
        self.assertEqual(self.graph.parse_vertices(), ["1", "2", "3", "4", "10"])

    def test_remove_vertex(self):
        with self.assertRaises(Exception):
            self.graph.remove_vertex(1)
        with self.assertRaises(Exception):
            self.graph.remove_vertex("10")
        self.graph.remove_vertex("2")
        self.assertEqual(self.graph.get_edge_cost("1", "3"), 3)
        self.assertEqual(self.graph.parse_outbound_edges("1"), ["3"])
        self.assertEqual(self.graph.parse_outbound_edges("3"), [])
        self.assertEqual(self.graph.parse_inbound_edges("1"), [])
        self.assertEqual(self.graph.parse_inbound_edges("3"), ["1"])
        self.assertEqual(self.graph.parse_vertices(), ["1", "3"])

    def test_add_edge(self):
        with self.assertRaises(Exception):
            self.graph.add_edge(1, "2", 10)
        with self.assertRaises(Exception):
            self.graph.add_edge("1", 2, 10)
        with self.assertRaises(Exception):
            self.graph.add_edge("1", "2", 10)
        with self.assertRaises(Exception):
            self.graph.add_edge("3", "2", "10")
        with self.assertRaises(Exception):
            self.graph.add_edge("3", "4", 10)
        self.graph.add_vertex("10")
        self.graph.add_vertex("4")
        self.graph.add_edge("10", "2", 1)
        self.graph.add_edge("10", "1", 10)
        self.graph.add_edge("1", "10", 3)
        self.graph.add_edge("2", "3", 4)
        self.assertEqual(self.graph.get_edge_cost("1", "3"), 3)
        self.assertEqual(self.graph.get_edge_cost("10", "2"), 1)
        self.assertEqual(self.graph.get_edge_cost("10", "1"), 10)
        self.assertEqual(self.graph.get_edge_cost("1", "10"), 3)
        self.assertEqual(self.graph.get_edge_cost("2", "3"), 4)
        self.assertEqual(self.graph.parse_outbound_edges("1"), ["2", "3", "10"])
        self.assertEqual(self.graph.parse_outbound_edges("2"), ["1", "3"])
        self.assertEqual(self.graph.parse_outbound_edges("3"), [])
        self.assertEqual(self.graph.parse_outbound_edges("4"), [])
        self.assertEqual(self.graph.parse_outbound_edges("10"), ["2", "1"])
        self.assertEqual(self.graph.parse_inbound_edges("1"), ["2", "10"])
        self.assertEqual(self.graph.parse_inbound_edges("2"), ["1", "10"])
        self.assertEqual(self.graph.parse_inbound_edges("3"), ["1", "2"])
        self.assertEqual(self.graph.parse_inbound_edges("10"), ["1"])
        self.assertEqual(self.graph.parse_inbound_edges("4"), [])
        self.assertEqual(self.graph.parse_vertices(), ["1", "2", "3", "10", "4"])

    def test_remove_edge(self):
        with self.assertRaises(Exception):
            self.graph.remove_edge(1, "2")
        with self.assertRaises(Exception):
            self.graph.remove_edge("1", 2)
        with self.assertRaises(Exception):
            self.graph.remove_edge("2", "3")
        with self.assertRaises(Exception):
            self.graph.remove_edge("3", "2")
        with self.assertRaises(Exception):
            self.graph.remove_edge("3", "4")
        self.graph.add_vertex("10")
        self.graph.add_vertex("4")
        self.graph.add_edge("10", "2", 1)
        self.graph.add_edge("10", "1", 10)
        self.graph.add_edge("1", "10", 3)
        self.graph.remove_edge("1", "2")
        self.graph.remove_edge("2", "1")
        self.graph.add_edge("2", "3", 4)
        self.assertEqual(self.graph.get_edge_cost("1", "3"), 3)
        self.assertEqual(self.graph.get_edge_cost("10", "2"), 1)
        self.assertEqual(self.graph.get_edge_cost("10", "1"), 10)
        self.assertEqual(self.graph.get_edge_cost("1", "10"), 3)
        self.assertEqual(self.graph.get_edge_cost("2", "3"), 4)
        self.assertEqual(self.graph.parse_outbound_edges("1"), ["3", "10"])
        self.assertEqual(self.graph.parse_outbound_edges("2"), ["3"])
        self.assertEqual(self.graph.parse_outbound_edges("3"), [])
        self.assertEqual(self.graph.parse_outbound_edges("4"), [])
        self.assertEqual(self.graph.parse_outbound_edges("10"), ["2", "1"])
        self.assertEqual(self.graph.parse_inbound_edges("1"), ["10"])
        self.assertEqual(self.graph.parse_inbound_edges("2"), ["10"])
        self.assertEqual(self.graph.parse_inbound_edges("3"), ["1", "2"])
        self.assertEqual(self.graph.parse_inbound_edges("10"), ["1"])
        self.assertEqual(self.graph.parse_inbound_edges("4"), [])
        self.assertEqual(self.graph.parse_vertices(), ["1", "2", "3", "10", "4"])
        self.graph.remove_vertex("1")
        self.assertEqual(self.graph.get_edge_cost("10", "2"), 1)
        self.assertEqual(self.graph.get_edge_cost("2", "3"), 4)
        self.assertEqual(self.graph.parse_outbound_edges("2"), ["3"])
        self.assertEqual(self.graph.parse_outbound_edges("3"), [])
        self.assertEqual(self.graph.parse_outbound_edges("4"), [])
        self.assertEqual(self.graph.parse_outbound_edges("10"), ["2"])
        self.assertEqual(self.graph.parse_inbound_edges("2"), ["10"])
        self.assertEqual(self.graph.parse_inbound_edges("3"), ["2"])
        self.assertEqual(self.graph.parse_inbound_edges("10"), [])
        self.assertEqual(self.graph.parse_inbound_edges("4"), [])
        self.assertEqual(self.graph.parse_vertices(), ["2", "3", "10", "4"])

    def test_eq(self):
        self.assertEqual(self.graph.copy(), self.graph)
        graph2 = self.graph.copy()
        graph2.add_vertex("4")
        self.assertNotEqual(graph2, self.graph)
        graph2.remove_vertex("1")
        self.assertNotEqual(graph2, self.graph)
        graph2.remove_vertex("4")
        self.assertNotEqual(graph2, self.graph)
        graph2.add_vertex("1")
        self.assertNotEqual(graph2, self.graph)
        graph2.add_edge("1", "2", 0)
        self.assertNotEqual(graph2, self.graph)
        graph2.add_edge("1", "3", 3)
        self.assertNotEqual(graph2, self.graph)
        graph2.add_edge("2", "1", 10)
        self.assertEqual(graph2, self.graph)
        self.graph.add_edge("1", "1", 0)
        graph2.add_edge("1", "1", 10)
        self.assertNotEqual(graph2, self.graph)

    def test_read_graph(self):
        with self.assertRaises(Exception):
            read_graph("aaaa")
        with self.assertRaises(Exception):
            read_graph("aaaa.modified")
        with self.assertRaises(Exception):
            read_graph("./test_files/fail1.txt")
        with self.assertRaises(Exception):
            read_graph("./test_files/fail2.txt")
        with self.assertRaises(Exception):
            read_graph("./test_files/fail3.txt")
        with self.assertRaises(Exception):
            read_graph("./test_files/fail1.modified.txt")
        with self.assertRaises(Exception):
            read_graph("./test_files/fail2.modified.txt")
        with self.assertRaises(Exception):
            read_graph("./test_files/fail3.modified.txt")
        with self.assertRaises(Exception):
            read_graph("./test_files/fail4.modified.txt")
        self.assertEqual(read_graph("./test_files/graph1.txt"), Graph(["0", "1", "2", "3", "4"], [
            ("0", "0", 1),
            ("0", "1", 7),
            ("1", "2", 2),
            ("2", "1", -1),
            ("1", "3", 8),
            ("2", "3", 5)
        ]))
        self.assertEqual(read_graph("./test_files/graph2.modified.txt"), Graph(["0", "1", "2", "3", "5"], [
            ("1", "5", 10),
            ("1", "2", 0),
            ("0", "3", 4),
            ("3", "5", 20),
            ("2", "1", 4),
            ("1", "1", 5)
        ]))

    def test_write_graph(self):
        with self.assertRaises(Exception):
            write_graph("aaaa", self.graph)
        with self.assertRaises(Exception):
            write_graph("aaaa.txt", self.graph)
        with self.assertRaises(Exception):
            write_graph("aaaa.modified.txt", "")
        write_graph("./test_files/aaaa.modified.txt", self.graph)
        self.assertEqual(read_graph("./test_files/aaaa.modified.txt"), self.graph)

    def test_random_graph(self):
        with self.assertRaises(Exception):
            random_graph(-10, 2)
        with self.assertRaises(Exception):
            random_graph(10, -2)
        with self.assertRaises(Exception):
            random_graph(6, 40)
        with self.assertRaises(Exception):
            random_graph("6", "2")
        graph = random_graph(6, 20)
        self.assertEqual(len(graph.parse_vertices()), 6)
        cnt_edges = 0
        for x in range(0, 6):
            cnt_edges += len(graph.parse_outbound_edges(str(x)))
        self.assertEqual(cnt_edges, 20)

    def test_floyd_warshall(self):
        self.graph.add_vertex("4")
        self.graph.add_edge("3", "4", 2)
        self.graph.add_edge("2", "4", 4)
        self.assertEqual(floyd_warshall(self.graph, "1", "4"), (4, ["1", "2", "4"]))
        self.assertEqual(floyd_warshall(self.graph, "4", "2"), (inf, []))
